modules/apps/29-fee/keeper/msg_server.go

```diff
@@ -0,0 +1,67 @@
+package keeper
+
+import (
+	"context"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+
+	"github.com/cosmos/ibc-go/v3/modules/apps/29-fee/types"
+	channeltypes "github.com/cosmos/ibc-go/v3/modules/core/04-channel/types"
+)
+
+var _ types.MsgServer = Keeper{}
+
+// RegisterCounterpartyAddress is called by the relayer on each channelEnd and allows them to specify their counterparty address before relaying
+// This ensures they will be properly compensated for forward relaying on the source chain since the destination chain must send back relayer's source address (counterparty address) in acknowledgement
+// This function may be called more than once by relayers, in which case, the previous counterparty address will be overwritten by the new counterparty address
+func (k Keeper) RegisterCounterpartyAddress(goCtx context.Context, msg *types.MsgRegisterCounterpartyAddress) (*types.MsgRegisterCounterpartyAddressResponse, error) {
```

Comment:
This looks to me like a global API, as I do not see the `channelEnd` being specified anywhere. But in a multi-chain setting, a relayer usually use the same wallet to relay from multiple chains. We cannot use the same counterparty address address for all channels that are relayed by the same relayer wallet, as the relayer's counterparty address for each chain is different.

Reply:
Would it be possible to register the forward relayer address on the source chain after this is registered on the destination chain? In theory, the source chain can also do the mapping with a merkle proof that the mapping exist on the destination chain.

If we can do the mapping, then there is no need for complicated version negotiation to enable fees. Each chain would be able to enable fee on any channel regardless of whether the counterparty chain has been upgraded, because the packet payloads do not need to be modified.

In the case that the source chain has fee enabled but the destination chain have not upgraded to support fee and the register counterparty address API, the only downside would just be that the escrowed receive fee is always refunded instead of being paid to the forward relayer.

On the other hand, if both chains are upgraded to support fee, the fee can still be asynchronously enabled on the source chain without it having to be enabled on the destination chain. In such case the relayer would still be able to register the counterparty and forward address on both chains, and then the source chain can still pay for the receive fee without having to modify the packet format.